File: tests/algorithms/transpiler/FS/electronics/electric_power.fs
// Generated 2025-08-07 15:46 +0700

exception Return
let mutable _nowSeed:int64 = 0L
let mutable _nowSeeded = false
let _initNow () =
    let s = System.Environment.GetEnvironmentVariable("MOCHI_NOW_SEED")
    if System.String.IsNullOrEmpty(s) |> not then
        match System.Int32.TryParse(s) with
        | true, v ->
            _nowSeed <- int64 v
            _nowSeeded <- true
        | _ -> ()
let _now () =
    if _nowSeeded then
        _nowSeed <- (_nowSeed * 1664525L + 1013904223L) % 2147483647L
        int _nowSeed
    else
        int (System.DateTime.UtcNow.Ticks % 2147483647L)

_initNow()
let rec _str v =
    let s = sprintf "%A" v
    s.Replace("[|", "[")
     .Replace("|]", "]")
     .Replace("; ", " ")
     .Replace(";", "")
     .Replace("\"", "")
type Result = {
    name: string
    value: float
}
let __bench_start = _now()
let __mem_start = System.GC.GetTotalMemory(true)
let rec absf (x: float) =
    let mutable __ret : float = Unchecked.defaultof<float>
    let mutable x = x
    try
        __ret <- if x < 0.0 then (-x) else x
        raise Return
        __ret
    with
        | Return -> __ret
let rec pow10 (n: int) =
    let mutable __ret : float = Unchecked.defaultof<float>
    let mutable n = n
    try
        let mutable p: float = 1.0
        let mutable i: int = 0
        while i < n do
            p <- p * 10.0
            i <- i + 1
        __ret <- p
        raise Return
        __ret
    with
        | Return -> __ret
let rec round_to (x: float) (n: int) =
    let mutable __ret : float = Unchecked.defaultof<float>
    let mutable x = x
    let mutable n = n
    try
        let m: float = pow10 (n)
        __ret <- (float (floor ((x * m) + 0.5))) / m
        raise Return
        __ret
    with
        | Return -> __ret
let rec electric_power (voltage: float) (current: float) (power: float) =
    let mutable __ret : Result = Unchecked.defaultof<Result>
    let mutable voltage = voltage
    let mutable current = current
    let mutable power = power
    try
        let mutable zeros: int = 0
        if voltage = 0.0 then
            zeros <- zeros + 1
        if current = 0.0 then
            zeros <- zeros + 1
        if power = 0.0 then
            zeros <- zeros + 1
        if zeros <> 1 then
            failwith ("Exactly one argument must be 0")
        else
            if power < 0.0 then
                failwith ("Power cannot be negative in any electrical/electronics system")
            else
                if voltage = 0.0 then
                    __ret <- { name = "voltage"; value = power / current }
                    raise Return
                else
                    if current = 0.0 then
                        __ret <- { name = "current"; value = power / voltage }
                        raise Return
                    else
                        if power = 0.0 then
                            let mutable p: float = absf (voltage * current)
                            __ret <- { name = "power"; value = round_to (p) (2) }
                            raise Return
                        else
                            failwith ("Unhandled case")
        __ret
    with
        | Return -> __ret
let rec str_result (r: Result) =
    let mutable __ret : string = Unchecked.defaultof<string>
    let mutable r = r
    try
        __ret <- ((("Result(name='" + (r.name)) + "', value=") + (_str (r.value))) + ")"
        raise Return
        __ret
    with
        | Return -> __ret
let r1: Result = electric_power (0.0) (2.0) (5.0)
printfn "%s" (str_result (r1))
let r2: Result = electric_power (2.0) (2.0) (0.0)
printfn "%s" (str_result (r2))
let r3: Result = electric_power (-2.0) (3.0) (0.0)
printfn "%s" (str_result (r3))
let r4: Result = electric_power (2.2) (2.2) (0.0)
printfn "%s" (str_result (r4))
let r5: Result = electric_power (2.0) (0.0) (6.0)
printfn "%s" (str_result (r5))
let __bench_end = _now()
let __mem_end = System.GC.GetTotalMemory(true)
printfn "{\n  \"duration_us\": %d,\n  \"memory_bytes\": %d,\n  \"name\": \"main\"\n}" ((__bench_end - __bench_start) / 1000) (__mem_end - __mem_start)

File: tests/algorithms/x/Pascal/electronics/electric_power.pas
{$mode objfpc}
program Main;
uses SysUtils, Math;
type Result = record
  name: string;
  value: real;
end;
var _nowSeed: int64 = 0;
var _nowSeeded: boolean = false;
procedure init_now();
var s: string; v: int64;
begin
  s := GetEnvironmentVariable('MOCHI_NOW_SEED');
  if s <> '' then begin
    Val(s, v);
    _nowSeed := v;
    _nowSeeded := true;
  end;
end;
function _now(): integer;
begin
  if _nowSeeded then begin
    _nowSeed := (_nowSeed * 1664525 + 1013904223) mod 2147483647;
    _now := _nowSeed;
  end else begin
    _now := Integer(GetTickCount64()*1000);
  end;
end;
function _bench_now(): int64;
begin
  _bench_now := GetTickCount64()*1000;
end;
function _mem(): int64;
var h: TFPCHeapStatus;
begin
  h := GetFPCHeapStatus;
  _mem := h.CurrHeapUsed;
end;
procedure panic(msg: string);
begin
  writeln(msg);
  halt(1);
end;
var
  bench_start_0: integer;
  bench_dur_0: integer;
  bench_mem_0: int64;
  bench_memdiff_0: int64;
  r1: Result;
  r2: Result;
  r3: Result;
  r4: Result;
  r5: Result;
  power: real;
  x: real;
  current: real;
  n: integer;
  r: Result;
  voltage: real;
function makeResult(name: string; value: real): Result; forward;
function absf(x: real): real; forward;
function pow10(n: integer): real; forward;
function round_to(x: real; n: integer): real; forward;
function electric_power(voltage: real; current: real; power: real): Result; forward;
function str_result(r: Result): string; forward;
function makeResult(name: string; value: real): Result;
begin
  Result.name := name;
  Result.value := value;
end;
function absf(x: real): real;
begin
  if x < 0 then begin
  exit(-x);
end;
  exit(x);
end;
function pow10(n: integer): real;
var
  pow10_p: real;
  pow10_i: integer;
begin
  pow10_p := 1;
  pow10_i := 0;
  while pow10_i < n do begin
  pow10_p := pow10_p * 10;
  pow10_i := pow10_i + 1;
end;
  exit(pow10_p);
end;
function round_to(x: real; n: integer): real;
var
  round_to_m: real;
begin
  round_to_m := pow10(n);
  exit(Floor((x * round_to_m) + 0.5) / round_to_m);
end;
function electric_power(voltage: real; current: real; power: real): Result;
var
  electric_power_zeros: integer;
  electric_power_p: real;
begin
  electric_power_zeros := 0;
  if voltage = 0 then begin
  electric_power_zeros := electric_power_zeros + 1;
end;
  if current = 0 then begin
  electric_power_zeros := electric_power_zeros + 1;
end;
  if power = 0 then begin
  electric_power_zeros := electric_power_zeros + 1;
end;
  if electric_power_zeros <> 1 then begin
  panic('Exactly one argument must be 0');
end else begin
  if power < 0 then begin
  panic('Power cannot be negative in any electrical/electronics system');
end else begin
  if voltage = 0 then begin
  exit(makeResult('voltage', power / current));
end else begin
  if current = 0 then begin
  exit(makeResult('current', power / voltage));
end else begin
  if power = 0 then begin
  electric_power_p := absf(voltage * current);
  exit(makeResult('power', round_to(electric_power_p, 2)));
end else begin
  panic('Unhandled case');
end;
end;
end;
end;
end;
end;
function str_result(r: Result): string;
begin
  exit(((('Result(name=''' + r.name) + ''', value=') + FloatToStr(r.value)) + ')');
end;
begin
  init_now();
  bench_mem_0 := _mem();
  bench_start_0 := _bench_now();
  r1 := electric_power(0, 2, 5);
  writeln(str_result(r1));
  r2 := electric_power(2, 2, 0);
  writeln(str_result(r2));
  r3 := electric_power(-2, 3, 0);
  writeln(str_result(r3));
  r4 := electric_power(2.2, 2.2, 0);
  writeln(str_result(r4));
  r5 := electric_power(2, 0, 6);
  writeln(str_result(r5));
  bench_memdiff_0 := _mem() - bench_mem_0;
  bench_dur_0 := (_bench_now() - bench_start_0) div 1000;
  writeln('{');
  writeln(('  "duration_us": ' + IntToStr(bench_dur_0)) + ',');
  writeln(('  "memory_bytes": ' + IntToStr(bench_memdiff_0)) + ',');
  writeln(('  "name": "' + 'main') + '"');
  writeln('}');
end.
